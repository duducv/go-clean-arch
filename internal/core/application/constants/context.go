package constants

type ContextKey string

const TXKey ContextKey = "tx"
