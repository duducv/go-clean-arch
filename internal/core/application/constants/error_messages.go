package constants

const (
	ErrInternal = "internal server error."
)
