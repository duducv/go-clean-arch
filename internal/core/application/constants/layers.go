package constants

type Layer string

const (
	ApplicationLayer Layer = "application"
	DomainLayer      Layer = "domain"
	InfraLayer       Layer = "infra"
)
