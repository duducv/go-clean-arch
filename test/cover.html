
<!DOCTYPE html>
<html>
	<head>
		<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
		<title>dto: Go Coverage Report</title>
		<style>
			body {
				background: black;
				color: rgb(80, 80, 80);
			}
			body, pre, #legend span {
				font-family: Menlo, monospace;
				font-weight: bold;
			}
			#topbar {
				background: black;
				position: fixed;
				top: 0; left: 0; right: 0;
				height: 42px;
				border-bottom: 1px solid rgb(80, 80, 80);
			}
			#content {
				margin-top: 50px;
			}
			#nav, #legend {
				float: left;
				margin-left: 10px;
			}
			#legend {
				margin-top: 12px;
			}
			#nav {
				margin-top: 10px;
			}
			#legend span {
				margin: 0 5px;
			}
			.cov0 { color: rgb(192, 0, 0) }
.cov1 { color: rgb(128, 128, 128) }
.cov2 { color: rgb(116, 140, 131) }
.cov3 { color: rgb(104, 152, 134) }
.cov4 { color: rgb(92, 164, 137) }
.cov5 { color: rgb(80, 176, 140) }
.cov6 { color: rgb(68, 188, 143) }
.cov7 { color: rgb(56, 200, 146) }
.cov8 { color: rgb(44, 212, 149) }
.cov9 { color: rgb(32, 224, 152) }
.cov10 { color: rgb(20, 236, 155) }

		</style>
	</head>
	<body>
		<div id="topbar">
			<div id="nav">
				<select id="files">
				
				<option value="file0">github.com/duducv/go-clean-arch/internal/core/application/dto/error_dto.go (100.0%)</option>
				
				<option value="file1">github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go (100.0%)</option>
				
				<option value="file2">github.com/duducv/go-clean-arch/internal/core/domain/entity/ticket.go (100.0%)</option>
				
				<option value="file3">github.com/duducv/go-clean-arch/internal/core/domain/entity/transaction.go (100.0%)</option>
				
				</select>
			</div>
			<div id="legend">
				<span>not tracked</span>
			
				<span class="cov0">not covered</span>
				<span class="cov8">covered</span>
			
			</div>
		</div>
		<div id="content">
		
		<pre class="file" id="file0" style="display: none">package dto

import "github.com/duducv/go-clean-arch/internal/core/application/constants"

type ErrorOutputLayer string

type ErrorOutput struct {
        Message    []string        `json:"message"`
        Raw        string          `json:"raw"`
        Layer      constants.Layer `json:"layer"`
        StatusCode int             `json:"statusCode"`
}

func NewErrorOutput(
        raw string,
        layer constants.Layer,
        statusCode int,
        message ...string,
) *ErrorOutput <span class="cov8" title="1">{
        return &amp;ErrorOutput{
                Message:    message,
                Raw:        raw,
                Layer:      layer,
                StatusCode: statusCode,
        }
}</span>
</pre>
		
		<pre class="file" id="file1" style="display: none">package usecases

import (
        "context"

        "github.com/duducv/go-clean-arch/internal/core/application/constants"
        "github.com/duducv/go-clean-arch/internal/core/application/dto"
        "github.com/duducv/go-clean-arch/internal/core/application/gateway"
        applicationrepository "github.com/duducv/go-clean-arch/internal/core/application/repository"
        "github.com/duducv/go-clean-arch/internal/core/domain/entity"
)

type TicketPurchaseUseCase struct {
        ticketRepository      entity.TicketRepository
        eventRepository       entity.EventRepository
        transactionRepository entity.TransactionRepository
        paymentGateway        gateway.PaymentGateway
        unitOfWorkRepository  applicationrepository.UnitOfWorkRepository
        trace                 applicationrepository.TracingRepository
}

func NewTicketPurchaseUseCase(
        ticketRepository entity.TicketRepository,
        eventRepository entity.EventRepository,
        transactionRepository entity.TransactionRepository,
        paymentGateway gateway.PaymentGateway,
        unitOfWorkRepository applicationrepository.UnitOfWorkRepository,
        trace applicationrepository.TracingRepository,
) TicketPurchaseUseCase <span class="cov8" title="1">{
        return TicketPurchaseUseCase{
                ticketRepository:      ticketRepository,
                eventRepository:       eventRepository,
                paymentGateway:        paymentGateway,
                transactionRepository: transactionRepository,
                unitOfWorkRepository:  unitOfWorkRepository,
                trace:                 trace,
        }
}</span>

func (usecase TicketPurchaseUseCase) Execute(
        ctx context.Context,
        input dto.TicketPurchaseInput,
) (*dto.TicketPurchaseOutput, *dto.ErrorOutput) <span class="cov8" title="1">{
        span, ctx := usecase.trace.StartSpan(ctx, "TicketPurchaseUseCase.Execute", string(constants.ApplicationLayer))
        defer span.End()
        ctx, err := usecase.unitOfWorkRepository.Begin(ctx)
        if err != nil </span><span class="cov8" title="1">{
                _ = usecase.unitOfWorkRepository.Rollback(ctx)
                return nil, dto.NewErrorOutput(err.Error(), constants.InfraLayer, 500, constants.ErrInternal)
        }</span>

        <span class="cov8" title="1">event, err := usecase.eventRepository.Get(ctx, input.EventId)
        if err != nil </span><span class="cov8" title="1">{
                _ = usecase.unitOfWorkRepository.Rollback(ctx)
                return nil, dto.NewErrorOutput(err.Error(), constants.InfraLayer, 500, constants.ErrInternal)
        }</span>

        <span class="cov8" title="1">ticket := entity.NewTicket(input.EventId, input.Email)
        err = usecase.ticketRepository.Save(ctx, ticket)
        if err != nil </span><span class="cov8" title="1">{
                _ = usecase.unitOfWorkRepository.Rollback(ctx)
                return nil, dto.NewErrorOutput(err.Error(), constants.InfraLayer, 500, constants.ErrInternal)
        }</span>

        <span class="cov8" title="1">transactionOutput, err := usecase.paymentGateway.CreateTransaction(ctx, dto.PaymentTransactionInput{
                Email: input.Email, CreditCardToken: input.CreditCardToken, Price: event.Price,
        })
        if err != nil </span><span class="cov8" title="1">{
                _ = usecase.unitOfWorkRepository.Rollback(ctx)
                return nil, dto.NewErrorOutput(err.Error(), constants.InfraLayer, 500, constants.ErrInternal)
        }</span>
        <span class="cov8" title="1">transaction := entity.NewTransaction(
                ticket.TicketID,
                event.EventID,
                transactionOutput.TID,
                event.Price,
                transactionOutput.Status,
        )
        err = usecase.transactionRepository.Save(ctx, transaction)
        if err != nil </span><span class="cov8" title="1">{
                _ = usecase.unitOfWorkRepository.Rollback(ctx)
                return nil, dto.NewErrorOutput(err.Error(), constants.InfraLayer, 500, constants.ErrInternal)
        }</span>

        <span class="cov8" title="1">if transactionOutput.Status == "approved" </span><span class="cov8" title="1">{
                ticket.Approve()
        }</span> else<span class="cov8" title="1"> {
                ticket.Cancel()
        }</span>

        <span class="cov8" title="1">err = usecase.ticketRepository.Update(ctx, ticket)
        if err != nil </span><span class="cov8" title="1">{
                _ = usecase.unitOfWorkRepository.Rollback(ctx)
                return nil, dto.NewErrorOutput(err.Error(), constants.InfraLayer, 500, constants.ErrInternal)
        }</span>

        <span class="cov8" title="1">err = usecase.unitOfWorkRepository.Commit(ctx)
        if err != nil </span><span class="cov8" title="1">{
                usecase.trace.CaptureError(ctx, err)
                _ = usecase.unitOfWorkRepository.Rollback(ctx)
                return nil, dto.NewErrorOutput(err.Error(), constants.InfraLayer, 500, constants.ErrInternal)
        }</span>

        <span class="cov8" title="1">return &amp;dto.TicketPurchaseOutput{
                TicketId: ticket.TicketID,
                Status:   string(ticket.Status),
                TID:      transactionOutput.TID,
                Price:    event.Price,
        }, nil</span>
}
</pre>
		
		<pre class="file" id="file2" style="display: none">package entity

import "github.com/lucsky/cuid"

type TicketStatus string

const (
        TicketStatusReserved TicketStatus = "reserved"
        TicketStatusApproved TicketStatus = "approved"
        TicketStatusCanceled TicketStatus = "canceled"
)

type Ticket struct {
        TicketID string
        EventID  string
        Email    string
        Status   TicketStatus
}

func NewTicket(
        eventID string,
        email string,
) Ticket <span class="cov8" title="1">{
        return Ticket{
                TicketID: cuid.New(),
                EventID:  eventID,
                Email:    email,
                Status:   TicketStatusReserved,
        }
}</span>

func (t *Ticket) Approve() <span class="cov8" title="1">{
        t.Status = TicketStatusApproved
}</span>

func (t *Ticket) Cancel() <span class="cov8" title="1">{
        t.Status = TicketStatusCanceled
}</span>
</pre>
		
		<pre class="file" id="file3" style="display: none">package entity

import "github.com/lucsky/cuid"

type Transaction struct {
        TransactionID string
        TicketID      string
        EventID       string
        TID           string
        Price         float64
        Status        string
}

func NewTransaction(
        ticketID string,
        eventID string,
        TID string,
        price float64,
        status string,
) Transaction <span class="cov8" title="1">{
        return Transaction{
                TransactionID: cuid.New(),
                TicketID:      ticketID,
                EventID:       eventID,
                TID:           TID,
                Price:         price,
                Status:        status,
        }
}</span>
</pre>
		
		</div>
	</body>
	<script>
	(function() {
		var files = document.getElementById('files');
		var visible;
		files.addEventListener('change', onChange, false);
		function select(part) {
			if (visible)
				visible.style.display = 'none';
			visible = document.getElementById(part);
			if (!visible)
				return;
			files.value = part;
			visible.style.display = 'block';
			location.hash = part;
		}
		function onChange() {
			select(files.value);
			window.scrollTo(0, 0);
		}
		if (location.hash != "") {
			select(location.hash.substr(1));
		}
		if (!visible) {
			select("file0");
		}
	})();
	</script>
</html>
