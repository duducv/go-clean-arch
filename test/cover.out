mode: set
github.com/duducv/go-clean-arch/internal/core/application/dto/error_dto.go:19.16,26.2 1 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:29.25,38.2 1 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:43.49,47.16 4 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:47.16,50.3 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:52.2,53.16 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:53.16,56.3 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:58.2,60.16 3 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:60.16,63.3 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:65.2,68.16 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:68.16,71.3 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:72.2,80.16 3 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:80.16,83.3 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:85.2,85.44 1 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:85.44,87.3 1 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:87.8,89.3 1 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:91.2,92.16 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:92.16,95.3 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:97.2,98.16 2 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:98.16,102.3 3 1
github.com/duducv/go-clean-arch/internal/core/application/usecases/ticket_purchase.go:104.2,109.8 1 1
github.com/duducv/go-clean-arch/internal/core/domain/entity/ticket.go:23.10,30.2 1 1
github.com/duducv/go-clean-arch/internal/core/domain/entity/ticket.go:32.28,34.2 1 1
github.com/duducv/go-clean-arch/internal/core/domain/entity/ticket.go:36.27,38.2 1 1
github.com/duducv/go-clean-arch/internal/core/domain/entity/transaction.go:20.15,29.2 1 1
